src/Agent.Sdk/SecretMasking/OssSecretMasker.cs

```diff
@@ -81,11 +106,132 @@ public void Dispose()
+                {
+                    break;
+                }
+            }
+        }
+
+        if (events < MaxTelemetryDetectionEvents && detectionData != null)
+        {
+            publishAction("SecretMaskerDetections", detectionData);
+            detectionData = null;
+        }
+
+        var overallData = new Dictionary<string, string>(GetOverallTelemetry());
+        if (_telemetry.Detections.Count > MaxTelemetryDetections)
+        {
+            overallData.Add("DetectionDataIsIncomplete", "true");
+        }
+
+        publishAction("SecretMasker", overallData);
+    }
+
+    private KeyValuePair<string, string>[] GetOverallTelemetry()
+    {
+        double elapsedMaskingTimeInMilliseconds = (double)_secretMasker.ElapsedMaskingTime / TimeSpan.TicksPerMillisecond;
+        return new KeyValuePair<string, string>[] {
+            new("Version", SecretMasker.Version.ToString()),
+            new("CharsScanned", _telemetry.CharsScanned.ToString(CultureInfo.InvariantCulture)),
+            new("StringsScanned", _telemetry.StringsScanned.ToString(CultureInfo.InvariantCulture)),
+            new("TotalDetections", _telemetry.TotalDetections.ToString(CultureInfo.InvariantCulture)),
+            new("UniqueCorrelatingIds", _telemetry.Detections.Count.ToString(CultureInfo.InvariantCulture)),
+            new("ElapsedMaskingTimeInMilliseconds", elapsedMaskingTimeInMilliseconds.ToString(CultureInfo.InvariantCulture)),
+        };
+    }
+
+    private sealed class Telemetry
+    {
+        private readonly ConcurrentDictionary<string, string> _detections;
+        private long _charsScanned;
+        private long _stringsScanned;
+        private long _totalDetections;
+
+        public IReadOnlyDictionary<string, string> Detections => _detections;
+        public long CharsScanned => _charsScanned;
+        public long StringsScanned => _stringsScanned;
+        public long TotalDetections => _totalDetections;
+        public readonly Action<Detection> DetectionAction;
+
+        public Telemetry()
+        {
+            _detections = new ConcurrentDictionary<string, string>();
+            DetectionAction = ProcessDetection;
+        }
+
+        private Telemetry(Telemetry copy)
+        {
+            _detections = new ConcurrentDictionary<string, string>(copy.Detections);
+            DetectionAction = ProcessDetection;
+
+            _charsScanned = copy.CharsScanned;
+            _stringsScanned = copy.StringsScanned;
+            _totalDetections = copy.TotalDetections;
+        }
+
+        public Telemetry Clone() => new(this);
+
+        public void ProcessInput(string input)
+        {
+            Interlocked.Add(ref _charsScanned, input.Length);
+            Interlocked.Increment(ref _stringsScanned);
+        }
+
+        private void ProcessDetection(Detection detection)
+        {
+            Interlocked.Increment(ref _totalDetections);
+
+            if (detection.CrossCompanyCorrelatingId != null)
```

Comment:
>CrossCompanyCorrelatingId [](http://example.com/codeflow?start=26&length=25)

OK, so obviously we are collecting and holding onto all observed c3ids for the scan session and I do think that's ok. In a different design, EnableTelemetry would have transported the callback, allowing you the option to batch stuff as you observed it. You'd still require a cache of observed c3ids to prevent double-firing, so it's not clear how much of a help this would be.

Still, as far as the pattern is concerned, if you received the telemetry callback in EnableTelemetry, you'd still have the option of maintaining the approach you've got here. That suggests that this pattern might be preferable if we expected a general use of this API (which we don't).

Reply:
I've now limited the memory use when the threshold is exceeded. 

I experimented with changing the pattern, but it makes it harder to reason about the threading and such IMHO. I suggest that we keep this pattern. It's possible that we'll want to do something similar in other consumers of this library. In that case, I think it would make sense to push the telemetry into security utilities as a feature and then it would be worth the expense of thinking about more cases that a generalization would afford. But as long as this is limited to the agent, I think it makes sense to keep it as simple as possible.